### ISSUE TITLE
`zeus start` exits with available commands waiting

### ISSUE THREAD

**Issue by keithpitty** [burke/zeus#570]
I've installed version 0.15.9 but it no longer starts.

![1 zeus-darwin-amd6 2016-07-01 10-36-53](https://cloud.githubusercontent.com/assets/11835/16508602/d5b8c1ca-3f77-11e6-8e37-fdda95525fc0.jpg)

Unfortunately I don't know where to start to try to debug this.

**Comment by keithpitty** [comment]
@metcalf I see you've been working on zeus lately. Thanks for that. Can you confirm that version 0.15.9 works for you? If so, I wonder what the critical difference is between our environments.

**Comment by metcalf** [comment]
I've yanked 0.15.9 from rubygems so that others don't suffer the same thing while we debug. This may be the same as https://github.com/burke/zeus/issues/567#issuecomment-229802574

Can you paste the full output of `zeus --log /dev/stderr start`?

**Comment by keithpitty** [comment]
```
cockatoo6 ~/tc/tc master[*] $ zeus --log /dev/stderr start
Starting Zeus server v0.15.9
[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── test_environment
    │   └── test_helper
    └── development_environment
        └── prerake

Available Commands: [waiting] [crashed] [ready]
zeus test (alias: rspec, testrb)
zeus generate (alias: g)
zeus destroy (alias: d)
zeus dbconsole
zeus rake
zeus runner (alias: r)
zeus console (alias: c)
zeus server (alias: s)
2016/07/01 10:44:10.580253 [/Users/andrew/code/go/src/github.com/burke/zeus/go/processtree/slavenode.go:193] boot/(no PID) in unbooted state
2016/07/01 10:44:10.583261 [/Users/andrew/code/go/src/github.com/burke/zeus/go/processtree/slavenode.go:373] boot/(no PID) running the root command now
2016/07/01 10:44:11.536564 [/Users/andrew/code/go/src/github.com/burke/zeus/go/processtree/slavenode.go:384] boot/(no PID) root process exited with an error before it could boot: exit status 1; output was: /Users/keithpitty/.gem/ruby/2.3.1/gems/zeus-0.15.9/lib/zeus/rails.rb:17:in `require': cannot load such file -- zeus (LoadError)
	from /Users/keithpitty/.gem/ruby/2.3.1/gems/zeus-0.15.9/lib/zeus/rails.rb:17:in `<top (required)>'
	from /Users/keithpitty/.rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in `require'
	from /Users/keithpitty/.rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in `rescue in require'
	from /Users/keithpitty/.rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:40:in `require'
	from /Users/keithpitty/tc/tc/custom_plan.rb:1:in `<top (required)>'
	from /Users/keithpitty/.rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Users/keithpitty/.rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
exit status 1
```

**Comment by metcalf** [comment]
We use Zeus pretty heavily at Stripe but have built a bunch of tooling on top of it. For example, we didn't encounter a critical bug in 0.15.8 because we pipe the output into other tools such that it doesn't use the TTY rendering mode you're seeing.

The biggest difference is probably that we're not a Rails shop so we're using Zeus exclusively as a Ruby runner.

**Comment by keithpitty** [comment]
@metcalf thanks for responding so quickly!

**Comment by keithpitty** [comment]
Thanks for explaining. I use Zeus for Rails projects, ideally in conjunction with RSpec but also to allow fast startup of a Rails server, console etc. I'm keen to get it working again, so if there's anything I can do to help, please let me know.

**Comment by metcalf** [comment]
I have to go meet my wife in a few minutes so I don't think I'll be able to get to this tonight. Can you roll back to 0.15.4 to get yourself unblocked in the meanwhile?

**Comment by keithpitty** [comment]
Sure. And thanks again for your prompt response. I'll let you know if I discover anything.

**Comment by metcalf** [comment]
I suspect switching things to use `require_relative` would fix whatever is going on here:
https://github.com/burke/zeus/pull/571

I haven't been able to repro the issue locally yet so I'm not sure if this is actually a fix.

**Comment by metcalf** [comment]
@latortuga: if you build and install #571, does it fix this issue for you?

**Comment by metcalf** [comment]
@keithpitty @latortuga: can one of you provide a minimal repro for this? The latest version runs just fine for me with rails:
```
cd $GOPATH/github.com/burke/zeus
make clean && make
gem install rubygem/pkg/*.gem

cd /tmp
rails new zeustest
cd zeustest
echo "gem 'zeus', '= 0.15.9'" >> Gemfile
cp $GOPATH/github.com/burke/zeus/examples/zeus.json .
bundle install
zeus --log /dev/stderr start
```

**Comment by metcalf** [comment]
(or @benjie on the minimal repro since you're following along too)

**Comment by keithpitty** [comment]
@metcalf I will do once I have my Go environment set up properly. Currently when I run `make` I get an error:

```
govendor fmt +local
make: govendor: No such file or directory
make: *** [fmt] Error 1
```

**Comment by metcalf** [comment]
@keithpitty: try `make dev_bootstrap`

**Comment by keithpitty** [comment]
@metcalf having made a few changes to your suggested command list, e.g. specifying the git source for zeus because 0.15.9 has been yanked, when I got as far as:

```
gem install $GOPATH/src/github.com/burke/zeus/rubygem/pkg/*.gem
```

I got the following error:

```
ERROR:  Could not find a valid gem '/Users/keithpitty/src/github.com/burke/zeus/rubygem/pkg/*.gem' (>= 0) in any repository
```

I now have to go out for the rest of the day but will check again in the evening.

**Comment by metcalf** [comment]
Git paths won't work with zeus unfortunately because the gem packages in the zeus binary. The `gem install` command (when it works) should convince bundler to use the local copy of 0.15.9 even though it can't find it on Rubygems. Is there anything in `ls $GOPATH/src/github.com/burke/zeus/rubygem/pkg/`? The `make clean && make` step should spit out a `.gem` file in there...

**Comment by metcalf** [comment]
I'm headed out of town for the holiday weekend and won't have my laptop but I'll help debug as possible here.

**Comment by benjie** [comment]
Here's what I did to get zeus installed on OSX:

```bash
brew install golang
echo 'export GOPATH="~/go"' >> ~/.profile_common
echo 'export PATH="$PATH:/usr/local/opt/go/libexec/bin"' >> ~/.profile_common
echo 'export PATH="$PATH:$GOPATH/bin"' >> ~/.profile_common
. ~/.profile_common
mkdir -p ~/go/src/github.com/burke
cd !$
git clone git@github.com/burke/zeus.git
cd zeus
rbenv install
make dev_bootstrap
make
cd ~/Documents/zeustest
rails new bob
cd bob
gem install $GOPATH/src/github.com/burke/zeus/rubygem/pkg/*.gem
```

---

Running `zeus --log /dev/stderr start` (under ruby 2.2.3) in a clean rails 5.0 produces:

```
zeus --log /dev/stderr start
Starting Zeus server v0.15.9
2016/07/02 12:04:06.587488 [/Users/benjiegillam/go/src/github.com/burke/zeus/go/processtree/slavenode.go:193] boot/(no PID) in unbooted state
2016/07/02 12:04:06.587659 [/Users/benjiegillam/go/src/github.com/burke/zeus/go/processtree/slavenode.go:373] boot/(no PID) running the root command now
[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── development_environment
    │   └── prerake
    └── test_environment
        └── test_helper

Available Commands: [waiting] [crashed] [ready]
zeus destroy (alias: d)
zeus dbconsole
zeus rake
zeus runner (alias: r)
zeus console (alias: c)
zeus server (alias: s)
zeus generate (alias: g)
zeus test (alias: rspec, testrb)
2016/07/02 12:04:06.884675 [/Users/benjiegillam/go/src/github.com/burke/zeus/go/processtree/slavenode.go:384] boot/(no PID) root process exited with an error before it could boot: exit status 1; output was: /Users/benjiegillam/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/zeus-0.15.9/lib/zeus/rails.rb:17:in `require': cannot load such file -- zeus (LoadError)
        from /Users/benjiegillam/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/zeus-0.15.9/lib/zeus/rails.rb:17:in `<top (required)>'
        from /Users/benjiegillam/.rbenv/versions/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `require'
        from /Users/benjiegillam/.rbenv/versions/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
        from /Users/benjiegillam/.rbenv/versions/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:39:in `require'
exit status 1
```

I had to actually Control-C to get back to the command prompt; pressing enter did nothing.

**Comment by benjie** [comment]
Weird; when I follow @metcalf's instructions (which seems the same as what I did) it does actually run.

```
➜ /tmp  rails new zeustest
cd       create
      create  README.rdoc
      create  Rakefile
      create  config.ru
      create  .gitignore
      create  Gemfile
      create  app
      create  app/assets/images/rails.png
      create  app/assets/javascripts/application.js
      create  app/assets/stylesheets/application.css
      create  app/controllers/application_controller.rb
      create  app/helpers/application_helper.rb
      create  app/mailers
      create  app/models
      create  app/views/layouts/application.html.erb
      create  app/mailers/.gitkeep
      create  app/models/.gitkeep
      create  config
      create  config/routes.rb
      create  config/application.rb
      create  config/environment.rb
      create  config/environments
      create  config/environments/development.rb
      create  config/environments/production.rb
      create  config/environments/test.rb
      create  config/initializers
      create  config/initializers/backtrace_silencers.rb
      create  config/initializers/inflections.rb
      create  config/initializers/mime_types.rb
      create  config/initializers/secret_token.rb
      create  config/initializers/session_store.rb
      create  config/initializers/wrap_parameters.rb
      create  config/locales
      create  config/locales/en.yml
      create  config/boot.rb
      create  config/database.yml
      create  db
      create  db/seeds.rb
      create  doc
      create  doc/README_FOR_APP
      create  lib
      create  lib/tasks
      create  lib/tasks/.gitkeep
      create  lib/assets
      create  lib/assets/.gitkeep
      create  log
      create  log/.gitkeep
      create  public
      create  public/404.html
      create  public/422.html
      create  public/500.html
      create  public/favicon.ico
      create  public/index.html
      create  public/robots.txt
      create  script
      create  script/rails
      create  test/fixtures
      create  test/fixtures/.gitkeep
      create  test/functional
      create  test/functional/.gitkeep
      create  test/integration
      create  test/integration/.gitkeep
      create  test/unit
      create  test/unit/.gitkeep
      create  test/performance/browsing_test.rb
      create  test/test_helper.rb
      create  tmp/cache
      create  tmp/cache/assets
      create  vendor/assets/javascripts
      create  vendor/assets/javascripts/.gitkeep
      create  vendor/assets/stylesheets
      create  vendor/assets/stylesheets/.gitkeep
      create  vendor/plugins
      create  vendor/plugins/.gitkeep
         run  bundle install
Fetching gem metadata from https://rubygems.org/...........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies..............
Using rake 11.2.2
Using i18n 0.7.0
Using multi_json 1.12.1
Using builder 3.0.4
Using erubis 2.7.0
Using journey 1.0.4
Using rack 1.4.7
Using hike 1.2.3
Using tilt 1.4.1
Using mime-types 1.25.1
Using polyglot 0.3.5
Using arel 3.0.3
Using tzinfo 0.3.50
Using bundler 1.11.2
Using coffee-script-source 1.10.0
Using execjs 2.7.0
Using json 1.8.3
Using thor 0.19.1
Using sass 3.4.22
Using sqlite3 1.3.11
Using activesupport 3.2.22.2
Using rack-cache 1.6.1
Using rack-test 0.6.3
Using rack-ssl 1.3.4
Using sprockets 2.2.3
Using treetop 1.4.15
Using coffee-script 2.4.1
Using uglifier 3.0.0
Using rdoc 3.12.2
Using activemodel 3.2.22.2
Using mail 2.5.4
Using actionpack 3.2.22.2
Using activerecord 3.2.22.2
Using activeresource 3.2.22.2
Using actionmailer 3.2.22.2
Using railties 3.2.22.2
Using coffee-rails 3.2.2
Using jquery-rails 3.1.4
Using rails 3.2.22.2
Using sass-rails 3.2.6
Bundle complete! 6 Gemfile dependencies, 40 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
➜ /tmp  cd zeustest
➜ /tmp/zeustest  echo "gem 'zeus', '= 0.15.9'" >> Gemfile
➜ /tmp/zeustest  echo "2.2" >> .ruby_version
➜ /tmp/zeustest  cp $GOPATH/src/github.com/burke/zeus/examples/zeus.json .
➜ /tmp/zeustest  gem install $GOPATH/src/github.com/burke/zeus/rubygem/pkg/*.gem
Successfully installed zeus-0.15.9
Parsing documentation for zeus-0.15.9
Done installing documentation for zeus after 0 seconds
1 gem installed
➜ /tmp/zeustest  bundle install
Resolving dependencies...
Using rake 11.2.2
Using i18n 0.7.0
Using multi_json 1.12.1
Using builder 3.0.4
Using erubis 2.7.0
Using journey 1.0.4
Using rack 1.4.7
Using hike 1.2.3
Using tilt 1.4.1
Using mime-types 1.25.1
Using polyglot 0.3.5
Using arel 3.0.3
Using tzinfo 0.3.50
Using bundler 1.11.2
Using coffee-script-source 1.10.0
Using execjs 2.7.0
Using json 1.8.3
Using thor 0.19.1
Using method_source 0.8.2
Using sass 3.4.22
Using sqlite3 1.3.11
Using activesupport 3.2.22.2
Using rack-cache 1.6.1
Using rack-test 0.6.3
Using rack-ssl 1.3.4
Using sprockets 2.2.3
Using treetop 1.4.15
Using coffee-script 2.4.1
Using uglifier 3.0.0
Using rdoc 3.12.2
Using zeus 0.15.9
Using activemodel 3.2.22.2
Using mail 2.5.4
Using actionpack 3.2.22.2
Using activerecord 3.2.22.2
Using activeresource 3.2.22.2
Using actionmailer 3.2.22.2
Using railties 3.2.22.2
Using coffee-rails 3.2.2
Using jquery-rails 3.1.4
Using rails 3.2.22.2
Using sass-rails 3.2.6
Bundle complete! 7 Gemfile dependencies, 42 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
➜ /tmp/zeustest  zeus --log /dev/stderr start
Starting Zeus server v0.15.9
2016/07/02 12:07:55.267265 [/Users/benjiegillam/go/src/github.com/burke/zeus/go/processtree/slavenode.go:193] boot/(no PID) in unbooted state
2016/07/02 12:07:55.267426 [/Users/benjiegillam/go/src/github.com/burke/zeus/go/processtree/slavenode.go:373] boot/(no PID) running the root command now
[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── development_environment
    │   └── prerake
[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── development_environment
    │   └── prerake
    └── test_environment
        └── test_helper

[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── development_environment
    │   └── prerake
    └── test_environment
        └── test_helper

Available Commands: [waiting] [crashed] [ready]
zeus generate (alias: g)
[ready] [crashed] [running] [connecting] [waiting]
boot
[ready] [crashed] [running] [connecting] [waiting]
boot
[ready] [crashed] [running] [connecting] [waiting]
boot
└── default_bundle
    ├── development_environment
    │   └── prerake
    └── test_environment
        └── test_helper

Available Commands: [waiting] [crashed] [ready]
zeus generate (alias: g)
zeus destroy (alias: d)
zeus dbconsole
zeus rake
zeus runner (alias: r)
zeus console (alias: c)
zeus server (alias: s)
zeus test (alias: rspec, testrb)
```

**Comment by benjie** [comment]
The difference was that I hadn't added zeus to my clean rails project's Gemfile. The moment I run `echo "gem 'zeus', '= 0.15.9'" >> Gemfile`, zeus now works.

**Comment by benjie** [comment]
👍  zeus 0.15.9 now running against my large rails project just fine; initial impressions are it seems a lot more stable than 0.15.4 - excellent work!

**Comment by keithpitty** [comment]
Isn't zeus supposed to be run without being specified in the `Gemfile`?

**Comment by keithpitty** [comment]
@metcalf if you remove `"gem 'zeus', '= 0.15.9'"` from your `Gemfile` do you see the error?

**Comment by metcalf** [comment]
Ahh, ok, I suspect it came out of this:
https://github.com/burke/zeus/pull/530/files

I think that PR seems reasonable but that the README needs to be updated to suggest using bundler. That might qualify as a backwards-incompatible change though. @kgrz @latortuga any thoughts here?

**Comment by metcalf** [comment]
ping @kgrz @latortuga on the proposed resolution, getting more pings to get a new version cut in: https://github.com/burke/zeus/issues/563

**Comment by kgrz** [comment]
Oops :( My bad in assuming that bundler is a prerequisite. I didn't think of a case where a rails app would be started without bundler. If this issue can be solved by reverting those commits, I'm all for it 👍 And as you mentioned, that change can go into the next major version once it's planned

I'm unable to recall the `Gem` api call that would load a particular gem without our code calling `gem 'name'`. I can check that tomorrow.

**Comment by metcalf** [comment]
OK, I've released 0.15.10 reverting #530 to fix this. Please reopen if it's still an issue!
